### PR TITLE
Update validators documentation for V2

### PR DESCRIPTION
**Pull request by tpdorsey**
## Change Summary

Updates usage guide to reflect Pydantic V2 features.

## Related issue number

Closes PYD-81
Closes PYD-170

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by tpdorsey** [comment]
please review

File: docs/usage/validators.md [hunk 1]
```diff
@@ -321,13 +345,18 @@ print(Model(a='invalid').a)
...
-## Validate always
+### Before and after validators
+
+In addition, you can execute code *before* or *after* field validation
+with the help of the `@field_validator` decorator `mode` parameter.
 
-**TODO this content is wrong!**
+`@field_validator`s can have `mode='before'`, in which case they're called before regular
+validation occurs (and are provided with the raw input data), or `mode='after'` (the default), in which case
+they're called after validation.
```
Comment (ybressler) [inline docs/usage/validators.md:355]: Worth mentioning the other modes, `"wrap", "plain"` and how they can be applied in annotated validators, but not decorated validators?

File: docs/usage/validators.md [hunk 2]
```diff
@@ -349,16 +378,13 @@ print(DemoModel(ts='2017-11-08T14:00'))
...
-Occasionally, you will want to use the same validator on multiple fields/models (e.g. to
-normalize some input data). The "naive" approach would be to write a separate function,
-then call it from multiple decorators.  Obviously, this entails a lot of repetition and
-boiler plate code. To circumvent this, the `allow_reuse` parameter has been added to
-`pydantic.validator` in **v1.2** (`False` by default):
+Occasionally, you will want to use the same validator on multiple fields or models (e.g. to
+normalize some input data).
+
+In Pydantic V2, you can simply define a validation function once, then reuse it with the
+`field_validator` function, specifying the field and validation function.
```
Comment (ybressler) [inline docs/usage/validators.md:387]: Perhaps worth demonstrating how to reuse validators in `mode=before|after`?

For ex:
```py

def first_char(v) -> Optional[str]:
    if isinstance(v, str) and len(v) > 1:
        return v[0]
    return v

class ModelA(BaseModel):
    a: str

    validate_a = field_validator("a", mode="before")(first_char)


class ModeB(BaseModel):
    b: str = "B"
   
    validate_b = field_validator("b", mode="after")(first_char)
```

File: docs/usage/validators.md [hunk 3]
```diff
@@ -388,88 +414,6 @@ assert jane_doe.name == 'Jane Doe'
...
 
-As it is obvious, repetition has been reduced and the models become again almost
-declarative.
-
-!!! tip
-    If you have a lot of fields that you want to validate, it usually makes sense to
-    define a help function with which you will avoid setting `allow_reuse=True` over and
-    over again.
-
-## Model validators
```
Comment (ybressler) [inline docs/usage/validators.md:399]: Just to verify, these changes just move stuff, yes? Or are there detailed changes?

**Review by ybressler — APPROVED** [review]
Docs look great. Thanks for adding these.

Q: If I wanted to add additional examples (in perhaps guide section), where would a good place to be for that?

File: docs/usage/validators.md (hunk 3 above)
Comment (tpdorsey) [reply docs/usage/validators.md:399]: These remove outdated documentation

Comment (ybressler) [reply docs/usage/validators.md:399]: Got it.

File: docs/usage/validators.md (hunk 1 above)
Comment (tpdorsey) [reply docs/usage/validators.md:355]: Adding a note and link to Annotated validators

File: docs/usage/validators.md (hunk 2 above)
Comment (tpdorsey) [reply docs/usage/validators.md:389]: Added this example

**Comment by tpdorsey** [comment]
@ybressler thanks for the suggestions.

You can definitely add examples, either to relevant sections or to a dedicated `## Examples` section if you like.

**Comment by samuelcolvin** [comment]
closing since replaced by #6695

**Pull request closed without merging**